src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,585 @@
+                    + std::to_string(last_modification_timestamp));
+            cache_iter->second.cache_controller->markInvalid();
+        }
+        else
+        {
+            // move the key to the list end
+            keys.splice(keys.end(), keys, cache_iter->second.key_iterator);
+            return {std::make_shared<LocalCachedFileReader>(cache_iter->second.cache_controller.get(), file_size), RemoteReadBufferCacheError::OK};
+        }
+    }
+
+    auto clear_ret = clearLocalCache();
+    cache_iter = caches.find(local_path);
+    if (cache_iter != caches.end())
+    {
+        if (cache_iter->second.cache_controller->isValid())
+        {
+            // move the key to the list end, this case should not happen?
+            keys.splice(keys.end(), keys, cache_iter->second.key_iterator);
+            return {std::make_shared<LocalCachedFileReader>(cache_iter->second.cache_controller.get(), file_size), RemoteReadBufferCacheError::OK};
+        }
+        else
+        {
+            // maybe someone is holding this file
+            return {nullptr, RemoteReadBufferCacheError::FILE_INVALID};
+        }
+    }
+
+    // reach the disk capacity limit
+    if (!clear_ret)
+    {
+        LOG_ERROR(log, "local cache is full, return nullptr");
+        return {nullptr, RemoteReadBufferCacheError::DISK_FULL};
+    }
+
+    std::filesystem::create_directories(local_path);
+
+    auto callback = [this](RemoteCacheController * cntrl) { this->total_size += cntrl->size(); };
+    auto cache_cntrl = std::make_shared<RemoteCacheController>(remote_file_meta, local_path, local_cache_bytes_read_before_flush, readbuffer, callback);
+    CacheCell cc;
+    cc.cache_controller = cache_cntrl;
+    cc.key_iterator = keys.insert(keys.end(), local_path);
+    caches[local_path] = cc;
+
+    return {std::make_shared<LocalCachedFileReader>(cache_cntrl.get(), file_size), RemoteReadBufferCacheError::OK};
+}
+
+bool RemoteReadBufferCache::clearLocalCache()
+{
+    for (auto it = keys.begin(); it != keys.end();)
+    {
+        auto cache_it = caches.find(*it);
+        auto cntrl = cache_it->second.cache_controller;
+        if (!cntrl->isValid() && cntrl->closable())
+        {
+            LOG_TRACE(log, "clear invalid cache: " + *it);
+            total_size = total_size > cache_it->second.cache_controller->size() ? total_size - cache_it->second.cache_controller->size() : 0;
+            cntrl->close();
+            it = keys.erase(it);
+            caches.erase(cache_it);
+        }
+        else
+            it++;
+    }
+    // clear closable cache from the list head
+    for (auto it = keys.begin(); it != keys.end();)
+    {
+        if (total_size < limit_size)
+            break;
+        auto cache_it = caches.find(*it);
+        if (cache_it == caches.end())
+        {
+            throw Exception("file not found in cache?" + *it, ErrorCodes::LOGICAL_ERROR);
+        }
+        if (cache_it->second.cache_controller->closable())
```

Comment:
Looks like some cache entries later than this one can become closable earlier, but you anyway now break from this cycle. To me it seems we can try to traverse futher.

Reply:
I don't realy get what is your meaning. 
It 's true that the later caches mybe closable, but we don't need to clear them. A closable cache means that there is no reader reading it at current, and it may be read soon later. So we just clear the  Least Frequently Used caches until enough disk space is released

Reply:
maybe you can take a look at the new implemnetation

Reply:
Ok, I will take a look at it a little later. BTW take a look at Common/LRUCache.h. May be you can reuse it?

Reply:
I know the LRU. but as you know, LRU drop keys on keys capacity not on value size

Reply:
You can manage it yourself. Look here https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/HashTable/LRUHashMap.h, it is in general idea the same as you wrote but in a good interface.